alvr/client_core/src/graphics/stream.rs

```diff
@@ -41,240 +38,276 @@ impl StreamRenderer {
-                        view: &render_objects.views_objects[view_idx].render_target
-                            [*swapchain_idx as usize],
-                        resolve_target: None,
-                        ops: wgpu::Operations {
-                            load: LoadOp::Clear(wgpu::Color::BLACK),
-                            store: StoreOp::Store,
-                        },
-                    })],
-                    ..Default::default()
-                });
-
-                render_pass.set_pipeline(&render_objects.pipeline);
-                render_pass.set_bind_group(
-                    0,
-                    &render_objects.views_objects[view_idx].bind_group,
-                    &[],
-                );
-                render_pass.draw(0..4, 0..1);
-            }
-
-            self.context.queue.submit(iter::once(encoder.finish()));
-        } else {
-            self.context.make_current();
-
-            #[cfg(all(target_os = "android", feature = "use-cpp"))]
-            super::opengl::renderStreamNative(hardware_buffer, swapchain_indices.as_ptr());
+        // if hardware_buffer is available copy stream to staging texture
+        if !hardware_buffer.is_null() {
+            self.staging_renderer.render(hardware_buffer);
         }
-    }
-}
 
-impl Drop for StreamRenderer {
-    fn drop(&mut self) {
-        self.context.make_current();
+        let mut encoder = self
+            .context
+            .device
+            .create_command_encoder(&Default::default());
 
-        #[cfg(all(target_os = "android", feature = "use-cpp"))]
-        if self.render_objects.is_none() {
-            unsafe { super::opengl::destroyStream() };
+        for (view_idx, swapchain_idx) in swapchain_indices.iter().enumerate() {
+            let mut render_pass = encoder.begin_render_pass(&RenderPassDescriptor {
+                label: None,
+                color_attachments: &[Some(RenderPassColorAttachment {
+                    view: &self.views_objects[view_idx].render_target[*swapchain_idx as usize],
+                    resolve_target: None,
+                    ops: wgpu::Operations {
+                        load: LoadOp::Clear(wgpu::Color::RED),
+                        store: StoreOp::Store,
+                    },
+                })],
+                ..Default::default()
+            });
+
+            render_pass.set_pipeline(&self.pipeline);
+            render_pass.set_push_constants(
+                ShaderStages::VERTEX_FRAGMENT,
+                0,
+                &(view_idx as u32).to_le_bytes(),
+            );
+            render_pass.set_bind_group(0, &self.views_objects[view_idx].bind_group, &[]);
+            render_pass.draw(0..4, 0..1);
         }
+
+        self.context.queue.submit(iter::once(encoder.finish()));
     }
 }
+
+pub fn foveated_encoding_shader_constants(
+    expanded_view_resolution: UVec2,
+    config: FoveatedEncodingConfig,
+) -> (UVec2, HashMap<String, f64>) {
```

Comment:
I only roughly checked all of this, it's just gonna have to be right

Reply:
Yes, I've tested all settings combinations, works perfectly like before